modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/dsl/AWS.scala

```diff
@@ -125,8 +128,39 @@ object AWS {
       }
     }
 
-    def readSqs(name: String, stop: Stream[F, Boolean]): Stream[F, Message[F, String]] =
-      SQS.readQueue(name, sqsVisibility.toSeconds.toInt, stop).map { case (msg, ack, extend) => Message(msg.body(), ack, extend) }
+
+    def readSqs(name: String, stop: Stream[F, Boolean]): Stream[F, String] = {
+      val awakePeriod: FiniteDuration = sqsVisibility - ExtendAllowance
+      SQS.readQueue(name, sqsVisibility.toSeconds.toInt, stop).flatMap { case (msg, ack, extend) =>
+        Stream.emit(msg.body())
+          .concurrently {
+            Stream.awakeEvery[F](awakePeriod).evalMap { _ =>
+              Logging[F].info(s"Approaching end of SQS message visibility. Extending visibility by $sqsVisibility.") *>
+              extend(sqsVisibility)
+            }
+            .handleErrorWith { t =>
+              Stream.eval(Logging[F].error(t)("Error extending SQS message visibility"))
+            }
+            .drain
+          }
+          .onFinalizeCase {
+            case ExitCase.Canceled =>
+              // The app is shutting down for a reason unrelated to processing this message.
+              // E.g. handling a SIGINT, or an exception was thrown processing a _different_ message, not this one.
+              ().pure[F]
+            case ExitCase.Error(t) =>
+              // This ExitCase means an exception was thrown upstream.
+              // But for this stream that can only mean when extending the visibility -- but we already handled all errors.
+              // So this case should never happen.
+              Logging[F].error(t)("Unexpected error waiting for SQS message to finalize")
+            case ExitCase.Completed =>
+              // This ExitCase means that the message was processed downstream.
+              // We get a ExitCase.Completed no matter if downstream ended in success or a raised exception.
+              // We ack the message, because in either case we don't want to read the SQS message again.
```

Comment:
So basically, a "typical" failure case in `Completed` would be Redshift/double-loading/S3 error, but much more often it would be success.

Reply:
I think this is less important, but what you said above about terminating `concurrently` branch is really amazing and we can mention both these cases.